core/src/memory.rs

```diff
@@ -130,138 +129,309 @@ pub fn vec_page_data_map_to_page_buf_map(
+    }
+    /// Constructs new page from raw page number with checks.
+    /// Returns error if page will contain bytes, with offsets bigger then u32::MAX.
+    fn new(num: u32) -> Result<Self, PageError> {
+        let page_begin = num
+            .checked_mul(Self::size())
+            .ok_or(PageError::OverflowU32MemorySize(num, Self::size()))?;
+        let last_byte_offset = Self::size() - 1;
+        // Check that the last page byte has index less or equal then u32::MAX
+        page_begin
+            .checked_add(last_byte_offset)
+            .ok_or(PageError::OverflowU32MemorySize(num, Self::size()))?;
+        // Now it is safe
+        unsafe { Ok(Self::new_unchecked(num)) }
+    }
+    /// Returns page zero byte offset.
+    fn offset(&self) -> u32 {
+        self.raw() * Self::size()
+    }
+    /// Returns page last byte offset.
+    fn end_offset(&self) -> u32 {
+        self.raw() * Self::size() + (Self::size() - 1)
+    }
+    /// Returns new page, which impls PageU32Size, and contains self zero byte.
+    fn to_page<PAGE: PageU32Size>(&self) -> PAGE {
+        PAGE::from_offset(self.offset())
+    }
+    /// Returns page which has number `page.raw() + raw`, with checks.
+    fn add_raw(&self, raw: u32) -> Result<Self, PageError> {
+        self.raw()
+            .checked_add(raw)
+            .map(Self::new)
+            .ok_or(PageError::AddOverflowU32(self.raw(), raw))?
+    }
+    /// Returns page which has number `page.raw() - raw`, with checks.
+    fn sub_raw(&self, raw: u32) -> Result<Self, PageError> {
+        self.raw()
+            .checked_sub(raw)
+            .map(Self::new)
+            .ok_or(PageError::SubOverflowU32(self.raw(), raw))?
+    }
+    /// Returns page which has number `page.raw() + other.raw()`, with checks.
+    fn add(&self, other: Self) -> Result<Self, PageError> {
+        self.add_raw(other.raw())
+    }
+    /// Returns page which has number `page.raw() - other.raw()`, with checks.
+    fn sub(&self, other: Self) -> Result<Self, PageError> {
+        self.sub_raw(other.raw())
+    }
+    /// Returns page which has number `page.raw() + 1`, with checks.
+    fn inc(&self) -> Result<Self, PageError> {
+        self.add_raw(1)
+    }
+    /// Returns page which has number `page.raw() - 1`, with checks.
+    fn dec(&self) -> Result<Self, PageError> {
+        self.sub_raw(1)
+    }
+    /// Aligns page zero byte and returns page which contains this byte.
+    /// Normally if `size % Self::size() == 0`,
+    /// then aligned byte is zero byte of the returned page.
+    fn align_down(&self, size: NonZeroU32) -> Self {
+        let size: u32 = size.into();
+        Self::from_offset((self.offset() / size) * size)
+    }
+    /// Returns page, which has zero byte offset == 0.
+    fn zero() -> Self {
+        unsafe { Self::new_unchecked(0) }
+    }
+    /// Returns iterator `self`..`self` + `count`.
+    fn iter_count(&self, count: Self) -> Result<PagesIter<Self>, PageError> {
```

Comment:
Cannot understand your proposal

Reply:
Each range type (e.g. `..N`, `N..=`, etc) implement `RangeBounds`. You can introduce method `fn iter_from_range<T: RangeBounds<Self>>(range: T) -> PagesIter<Self>` instead of all of these

Reply:
How you will do iter_count ??

Reply:
And also we have two iterators types